### PR TITLE
Add SSA conditions helpers in internal/utils

### PR DESCRIPTION
**Pull request by fwiesel**
Foundation for migrating all controllers to Server-Side Apply of the status subresource.

Introduces two helpers in `internal/utils/conditions.go`:

- `ConditionFromStatus`: converts a single `metav1.Condition` (as stored in a status subresource) to a `ConditionApplyConfiguration` suitable for SSA. `ObservedGeneration` is preserved so gating logic (e.g. the taint controller's skip-check) keeps working.
- `SetApplyConfigurationStatusCondition`: mirrors the semantics of `meta.SetStatusCondition` on the apply-configuration slice: append or update by `Type`, refresh `LastTransitionTime` only when `Status` changes, guard against nil/empty `Type` to prevent panics.

Together they enable per-condition seeding so each controller claims field ownership only over the conditions it actually manages, without disturbing conditions owned by other field managers on Apply.

## Summary by CodeRabbit

* **New Features**
  * Added utilities to convert status conditions into apply-configuration form.
  * Added functionality to add or update conditions in-place by type, with correct handling of timestamps, observed generation, and no-op behavior when inputs are invalid or unchanged.

* **Tests**
  * Added comprehensive test coverage for condition conversion and in-place updates, including append vs. update flows, transition time behavior, and edge cases for missing/empty fields.

### CODE REVIEW

**Comment by fwiesel** [comment]
Consumer PRs that depend on this foundation:

- #336 EvictionReconciler
- #337 HypervisorController
- #338 HypervisorTaintController
- #339 HypervisorMaintenanceController (status + eviction creation)
- #340 OnboardingController
- #341 HypervisorOffboardingReconciler
- #342 AggregatesController
- #343 TraitsController
- #344 ready.Controller
- #345 NodeCertificateController

Final cleanup that removes `internal/utils/status_patch.go` once all callers are gone: #346

File: internal/utils/conditions.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+			now := metav1.Now()
+			newCondition.LastTransitionTime = &now
+		}
+		*conditions = append(*conditions, newCondition)
+		return true
+	}
+
+	// Existing condition: update fields, handle LastTransitionTime
+	statusChanged := (existing.Status == nil) != (newCondition.Status == nil) ||
+		(existing.Status != nil && *existing.Status != *newCondition.Status)
```
Comment (joker-at-work) [inline internal/utils/conditions.go:86]: Why can't this use `strChanged`? If I [read it correctly](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#ConditionStatus) `ConditionStatus` is a type of string.

Comment (joker-at-work) [reply internal/utils/conditions.go:86]: Ok, it's because golang doesn't see it as string anymore if it's a named string type.
AI suggests to "introduce a generic ptrChanged[T comparable] helper" to replace all 3 needed compare helpers. In any case, I got my answer.

**Review by joker-at-work — APPROVED** [review]

**Review by notandy — COMMENTED** [review]
I was thinking that we could extend v1.Hypervisor{} with a unexported field that holds (an optional) SSA `HypervisorStatusApplyConfiguration`. This way we wouldn't need to pass it along every other function and could check for changes (against it's own self.Status) on demand when needed, instead of unconditionally.

**Comment by fwiesel** [comment]
> I was thinking that we could extend v1.Hypervisor{} with a unexported field that holds (an optional) SSA HypervisorStatusApplyConfiguration. This way we wouldn't need to pass it along every other function and could check for changes (against it's own self.Status) on demand when needed, instead of unconditionally.

Not sure, if I got your point correctly:  In my mind, SSA make it easier to just declare the desired "state". No need to check, if something needs to get added/removed, changed or not.  That being said, that is the ideal state, and often you need the information.

The code now is more or less just replacing the patch with SSA apply and just for the status and intentionally not refactored, as the delta is already huge.

I think, there is still a lot of room of improvement, it was just the "smallest unit of change" I thought sensible.

**Comment by notandy** [comment]
> > I was thinking that we could extend v1.Hypervisor{} with a unexported field that holds (an optional) SSA HypervisorStatusApplyConfiguration. This way we wouldn't need to pass it along every other function and could check for changes (against it's own self.Status) on demand when needed, instead of unconditionally.
> 
> Not sure, if I got your point correctly: In my mind, SSA make it easier to just declare the desired "state". No need to check, if something needs to get added/removed, changed or not. That being said, that is the ideal state, and often you need the information.
> 
> The code now is more or less just replacing the patch with SSA apply and just for the status and intentionally not refactored, as the delta is already huge.
> 
> I think, there is still a lot of room of improvement, it was just the "smallest unit of change" I thought sensible.

Exactly, the code is just replacing SSA apply without really caring about SSA, and forcing every callee to accept an additional field in their signature. The request makes sense because it would reduce the boilerplate introduced by the switch to SSA. Although this PR is large, you're asking to cling on some old behaviour while introducing a massive change - and requiring to follow up with other big changes to make it "nicer". Not sure I see the logic in this.

**Comment by fwiesel** [comment]
>  Not sure I see the logic in this.

A proper refactoring would have touched effectively all lines of codes. 

I think, now I understand:
>  and forcing every callee to accept an additional field in their signature
>>  I was thinking that we could extend v1.Hypervisor{} with a unexported field that holds (an optional) SSA HypervisorStatusApplyConfiguration.

I do not think, that is a good idea. Preserving the signature of a function only makes sense to me if there are external users, and an API need to be kept. I do not understand the need to enforce that in private functions. Putting it inside another object simply hides that the function needs that information. 

I do see that there is a duplication between `Hypervisor` and `HypervisorApplyConfiguration`, but I think, it is part of how ApplyConfigurations works The first is the current state, and all mandatory fields need to be set, and come from the API-server, while *ApplyConfiguration` allows all fields to be optional, as multiple controllers together can compose an object.
In an ideal world, we wouldn't have that "duplication", but that is outside of our control, unless we want to forgo SSA completely.

**Comment by notandy** [comment]
> I do see that there is a duplication between `Hypervisor` and `HypervisorApplyConfiguration`

It's less about not changing the signature at all, but that you introduce a behavioural change. Before SSA, the `Hypervisor` object was modified and then (eventually) applied. If you change the meaning of your private functions to use `Hypervisor` as a reference and `HypervisorApplyConfiguration` as the Changeset, it should be clear how the function is used and behaving. e.g. Hypervisor shouldn't be a reference anymore. And maybe you could change the support function to return an array of `ConditionApplyConfiguration`s?

> In an ideal world, we wouldn't have that "duplication", but that is outside of our control, unless we want to forgo SSA completely.

Let's keep the hypothetical out of this discussion, there are multiple ways to achieve the same goal, let's find the best one.

**Review by notandy — APPROVED** [review]

**Review by joker-at-work — APPROVED** [review]
